### ISSUE TITLE
Troubleshoot timeout issues for VSA Caregiver connection to Salesforce/CARMA

### ISSUE THREAD

**Issue by jhouse-solvd** [department-of-veterans-affairs/va.gov-team#19503]
## Description
The  #vsa-caregiver team has been seeing an increasing amount of timeout errors occurring with the connection to Salesforce/CARMA. Although these are timeout errors, the data is going through and being processed successfully. This is resulting in false negatives for the user and lots of records in salesforce without the required attachment (since processing ended on the timeout error).
The timeout is happening at 60 seconds and CARMA confirmed that this is much more time than what’s needed to process the data. Since this looks like a networking issue, we are hoping someone on the operations/infrastructure side can investigate and diagnose the issue.

## Background/context/resources
Related vsa-caregiver ticket: https://github.com/department-of-veterans-affairs/va.gov-team/issues/18555 (<-- lots of great previous investigation and troubleshooting in this ticket)
Related platform thread: https://dsva.slack.com/archives/CBU0KDSB1/p1611938061012300
Sentry Errors:
http://sentry.vfs.va.gov/organizations/vsp/issues/6552/events/latest/?environment=pr[…]%3Aunresolved+controller_name%3Acaregivers_assistance_claims
http://sentry.vfs.va.gov/organizations/vsp/issues/5207/events/f518b5becfab4d76b2e877[…]%3Aunresolved+controller_name%3Acaregivers_assistance_claims

## Technical notes
_Notes around work that is happening, if applicable_

---
## Tasks
- [x] Narrow down the possibilities for where this timeout might be triggerd
- [x] Troubleshoot and determine path for remediation
- [x] Discuss with VSA Caregiver team as needed to validate assumptions and or resolve the issue
- [ ] Resolve the issue (if within Ops ability to do so) 

## Definition of Done
- [ ] _What is the expected outcome of this issue?_

---
### Reminders
- [X] Please attach your team label and any other appropriate label(s) 
- [X] Please attach the needs grooming tag if needed
- [X] Please connect to an epic

**Comment by jhouse-solvd** [comment]
@dginther will review the information provided by the vsa-caregiver team in https://github.com/department-of-veterans-affairs/va.gov-team/issues/18555 and help narrow down a good next step

**Comment by jhouse-solvd** [comment]
@kevinmirc - Okay, thank you for the update! Let us know if we need to re-open this. Closing for now.

**Comment by jhouse-solvd** [comment]
@kevinmirc / @mchelen: 

some additional technical bits to consider: 

- There is nothing in DevOps land that would arbitrarily would slow / throttle the connection, so the database processing time is still something to consider on the UX side of the application (ie background the process so the user can continue?)
- The default fwdproxy timeout is 1m (aka 60 seconds) 
- In at least one case (specifically the connection through fwdproxy for EVSS found in fwdproxy-vagov-prod.yml) the "server_timeout" setting has been set explicitly configured at 6m (presumably to account for longer processing / transfer time)
- It's possible that a higher timeout value for the network connection through fwdproxy will be needed, depending on how you want to proceed

Thought this might be useful as you approach solving this issue.

**Comment by kevinmirc** [comment]
Thank you, @jhouse-solvd for the additional information. [I was told](https://dsva.slack.com/archives/CBU0KDSB1/p1611943669054300?thread_ts=1611938061.012300&cid=CBU0KDSB1) that the fwdproxy timeout had a default of 360 seconds. Can we have someone confirm that salesforce is using a fwdproxy and what the timeout for the proxy is set to?

**Comment by kevinmirc** [comment]
@jbritt1 confirmed on [slack](https://dsva.slack.com/archives/CBU0KDSB1/p1614181493201900?thread_ts=1611938061.012300&cid=CBU0KDSB1) that the fwdproxy timeout is 360 seconds.

**Comment by jhouse-solvd** [comment]
@kevinmirc - Here's some lengthy information, but worth considering...

There are a few different HA proxy configuration settings to consider for forward proxy, with respect to timeout of connections, aside from what's configured at the AWS load balancer level: 

The defaults specified in [this config](https://github.com/department-of-veterans-affairs/devops/blob/master/ansible/deployment/config/fwdproxy/haproxy.conf.j2) are (with my simplified explanation):

timeout http-request    10s - this is the window for the request to be made and a response to be received
timeout client          1m - this is the window during which we expect the client to be sending data
timeout server          1m - this is the window during which we expect the backend to be sending data
et al

However, specific backends can also be configured to override those defaults, and as you can see [in this example for EVSS](https://github.com/department-of-veterans-affairs/devops/blob/e2300806aca64f1404765ed04c018136688361a6/ansible/deployment/config/fwdproxy-vagov-prod.yml#L28), a server_timeout value of 6m has been specified, presumably to allow for longer processing time between client and server. In this case is that the client is a the FE app, and the server is www.ebenefits.va.gov as specified in the "host" declaration.

In the case of the salesforce backend connection to va.my.salesforce.com (specified [here](https://github.com/department-of-veterans-affairs/devops/blob/e2300806aca64f1404765ed04c018136688361a6/ansible/deployment/config/fwdproxy-vagov-prod.yml#L184)), "server_timeout" has not been declaratively set, which means that the default value of 1m would come into play. 

Based on the information in this ticket, it looks like there is a legitimate reason for processing times of > 1m, and the vsa-caregiver team would like to experiment with increasing the timeout for those connections. In speaking with Demian and the Operations team, it seems that the "server_timeout" value might also need to be increased for the salesforce backend configuration, similar to what's been done for EVSS.

TL;DR - We recommend submitting a PR to explicitly set the "server_timeout" value to 120s for the salesforce backend configuration through fwdproxy (in STAGING first if it's possible to test there), in addition to any application-level settings that you may have submitted for review.

Hopefully this is fairly clear, but happy to review with you more specifically. :)

**Comment by kevinmirc** [comment]
@jhouse-solvd, Thanks for providing this info. I'll go ahead and open up a pull request as you suggested.

I'm not so sure we'll be able to do a "real" test (a 60+ second submission). What I can do is apply the changes to the staging and ensure that applications on our form can still go through as expected. Then apply the same changes to production and monitor logs closely to ensure there was no impact to our form. I think this would be sufficient.

**Comment by jhouse-solvd** [comment]
@kevinmirc - awesome. let us know how your testing goes. I've re-opened the ticket, and adding the label "awaiting-feedback"

**Comment by kevinmirc** [comment]
Hi all,

On Friday (February 26, 2021), we did some testing on a RI to view the expected UI when processing times are longer than 60 seconds ([PR here](https://github.com/department-of-veterans-affairs/vets-website/pull/16166)).

To test this we updated the controller, in the RI, to pause for 90 seconds: 
```rb
# app/controllers/v0/caregivers_assistance_controller.rb
module V0
  class CaregiversAssistanceClaimsController < ApplicationController
    def create
      sleep 90.seconds
    end
  end
end
```

then sent the form submission request and watched the UI.

I noticed that after *60* seconds, the request timed out by a server called "openresty".

<img width="812" alt="Screen Shot 2021-02-26 at 12 08 43 PM" src="https://user-images.githubusercontent.com/7391284/109543992-fe972980-7a94-11eb-8753-c94dd602de6b.png">

I'm assuming this is the reverse proxy timing out. If that's the case, I think as part our epic to [Improve user experience of 10-10CG - Long submission times](https://app.zenhub.com/workspaces/vsa---caregiver-5fff0cfd1462b6000e320fc7/issues/department-of-veterans-affairs/va.gov-team/19321), we may need to consider extending the reverse proxy's timeout setting.

TL;DR:
- Are you able to confirm that the rev-proxy has a 60-second timeout?
- Is it possible to extend the timeout of the rev-proxy for this request?
- What are the steps involved in extending this timeout?

**Comment by patrickvinograd** [comment]
The preferred, but also more involved, solution, would be to make the processing of this request asynchronous, so that the initial request can be submitted quickly, it spawns a background worker to submit to CARMA or whatever, and then there's an API endpoint to poll for status.

There are several endpoints in vets-api that have  been refactored to follow this approach after discovering high latency/processing time in upstream systems.

**Comment by kevinmirc** [comment]
@patrickvinograd Our followup solution is to restructure the submission as you described but we also want to understand if extending the timeout is worth doing in the short-term to alleviate errors in the meantime.

Regardless, we still need to have a better understanding of how the revproxy is affecting our application today. Based on testing we did in the RI, it looks like the reverse proxy is sending a response to the UI **before** the API is. So we would need understand if this is happening in production too. Are you able to point me to the revproxy settings to confirm that it's 60s in production also?

**Comment by kevinmirc** [comment]
@jhouse-solvd This ticket is ready to close since our team now has everything we need to make outbound (API => CARMA) requests and change the timeout of those requests. I don't think there is anything left here for Ops.

Thanks again everyone for the guidance.

**Comment by jhouse-solvd** [comment]
@kevinmirc - thanks so much and glad to hear it. closing.
